community/contributing/pr_workflow.rst

```diff
@@ -187,42 +197,55 @@ To ensure there won't be conflicts between the feature you develop and the
-
-Therefore, the command that you should (almost) always use is:
-
 ::
 
     $ git pull --rebase upstream master
 
-
-If you have already pushed the merge commit without using ``rebase``, or
-have made any other changes that have resulted in undesired history, you may
-use a hard reset to revert to a specific commit and try again:
-
-::
-
-    $ git reset --hard [The ID of the last desired commit]
-
-Once you have done this, you may run ``--rebase`` to merge master correctly.
-
-If you have already pushed the wrong commits to your remote branch,
-you will have to force push by using ``git push --force``.
-
-.. warning:: ``git reset --hard`` can be a dangerous operation, especially
-            if you have untracked or uncommitted changes. However, if
-            you have committed changes that you reset using ``git reset --hard``,
-            you may still be able to recover them by resetting to a commit ID
-            found with the ``git reflog`` command.
+ The ``--rebase`` argument will ensure that any local changes that you committed
+will be re-applied *on top* of the pulled branch, which is usually what we want
+in our PR workflow. This way, when you open a pull request, your own commits will
+be the only difference with the upstream ``master`` branch.
+
+While rebasing, conflicts may arise if your commits modified code that has been
+changed in the upstream branch in the meantime. If that happens, Git will stop at
+the conflicting commit and will ask you to resolve the conflicts. You can do so
+with any text editor, then stage the changes (more on that later), and proceed with
+``git rebase --continue``. Repeat the operation if later commits have conflicts too,
+until the rebase operation completes.
+
+If you're unsure about what is going on during a rebase and you panick (no worry,
+we all do the first few times), you can abort the rebase with ``git rebase --abort``.
+You will then be back to the original state of your branch before calling
+``git pull --rebase``.
+
+.. note:: If you omit the ``--rebase`` argument, you will instead create a merge
+          commit which tells Git what to make of the two distinct branches. If any
+          conflicts arise, they would be resolved all at once via this merge commit.
+
+          While this is a valid workflow and the default behavior of ``git pull``,
+          merge commits within PRs in our PR workflow. We only use them when merging
+          PRs into the upstream branch.
+
+          The philosophy is that a PR should represent the final stage of the changes
+          made to the codebase, and we are not interested in mistakes and fixes that
+          would have been done in intermediate stages before merging.
+          Git gives us great tools to "rewrite the history" and make it as if we got
+          things right the first time, and we're happy to use it to ensure that
+          changes are easy to review and understand long after they have been merged.
```

Comment:
```suggestion
```

I suggest removing this sentence; the previous one does enough to convey the reasoning, IMO. As is, this sentence does encourage "rewriting history" a little too much in my opinion, and would need an added caveat that while that's fine in cases such as this were said history doesn't go back that far and isn't public yet, it needs caution (if done at all) in other cases, because it can really mess things up. But... that then becomes pretty long :)